broadinstitute/firecloud-orchestration
# GAWB-1644: copy ontology parent-node information into Library

Warning: PO has not seen this yet; it is possible functionality will need to change.

When indexing datasets into Library's ElasticSearch, if the dataset is tagged with an ontology node id for the "indication" field, query consent-ontology for that node's parents. Then, save the parent information in Library's ES.

By having the parent information present in ES, we enable ontology-aware search.

To test:
* catalog a dataset. While cataloging, make a note of the indication's node id - for instance, `http://purl.obolibrary.org/obo/DOID_1485` for cystic fibrosis.
* publish the dataset.
* look up the node id using consent-ontology's search API: https://consent-ontology.dsde-dev.broadinstitute.org/swagger/#!/Search/get_search, and make note of the parents' labels. For instance, monogenic disease is a parent of cystic fibrosis.
* go to Library in the UI and enter a parent label into the main search box, and see that your dataset is returned in search results.

Notes:
* I have explicitly chosen to NOT make parent-node definitions or synonyms searchable. This is a precision/recall tradeoff - I believe making these searchable will decrease precision and increase recall by too much.
* I have explicitly chosen to NOT make parent-node labels available in autocomplete suggestions. This is in order to not clutter the suggestions.
* I am not utilizing the `order` value of the parent to assist with relevancy. This was a stretch goal; I'll handle that in a future story.

There's a ton of commits in this PR, sorry!

  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
